File: python/day16.py
import unittest
from utilities import read_input
from collections import defaultdict
from typing import List, Dict, Tuple

test_input = """class: 1-3 or 5-7
row: 6-11 or 33-44
seat: 13-40 or 45-50

your ticket:
7,1,14

nearby tickets:
7,3,47
40,4,50
55,2,20
38,6,12"""

test_input2 = """class: 0-1 or 4-19
row: 0-5 or 8-19
seat: 0-13 or 16-19

your ticket:
11,12,13

nearby tickets:
3,9,18
15,1,5
5,14,9"""


class FieldRule:
    def __init__(self, rule: str):
        rule: List[str] = rule.split(' or ')
        self.first_range: Tuple[int, int] = self._read_range(rule[0])
        self.second_range: Tuple[int, int] = self._read_range(rule[1])

    def _read_range(self, range: str) -> Tuple[int, int]:
        range: List[str, str] = range.split('-')
        return int(range[0]), int(range[1])

    def validate(self, value: int) -> bool:
        return self.first_range[0] <= value <= self.first_range[1] or\
            self.second_range[0] <= value <= self.second_range[1]


class TicketRule:
    def __init__(self, rules: str):
        self.rules: Dict[str, FieldRule] = {}
        self.rule_index: Dict[str, int] = {}
        self.valid_tickets: List[List[int]] = []
        for rule in rules.splitlines():
            self.set_rule(rule)

    def set_rule(self, rule: str):
        parts: List[str] = rule.split(': ')
        field: str = parts[0]
        self.rules[field] = FieldRule(parts[1])

    def _validate_value(self, value: int) -> bool:
        return any(map(lambda rule: rule.validate(value), self.rules.values()))

    def count_invalid(self, ticket: List[int]) -> int:
        invalid = 0
        valid = True
        for value in ticket:
            if not self._validate_value(value):
                valid = False
                invalid += value
        if valid:
            self.valid_tickets.append(ticket)
        return invalid

    def validate_rule(self, field: str, ticket_idx: int) -> bool:
        rule: FieldRule = self.rules[field]
        for ticket in self.valid_tickets:
            if not rule.validate(ticket[ticket_idx]):
                return False
        return True

    def detect_index(self):
        # Find the possible indices for each field
        possile_idx: Dict[str, List[int]] = defaultdict(list)
        for field in self.rules:
            for i in range(len(self.rules)):
                if self.validate_rule(field, i):
                    possile_idx[field].append(i)

        # Sort by len of possible choices to help find faster.
        possible_idx_len = [(len(possile_idx[field]), field)
                            for field in self.rules]
        possible_idx_len.sort()

        used_indices = [False] * len(self.rules)

        def helper(field_idx):
            if field_idx == len(self.rules):
                return True

            field = possible_idx_len[field_idx][1]
            for idx in possile_idx[field]:
                if not used_indices[idx]:
                    used_indices[idx] = True
                    if helper(field_idx + 1):
                        self.rule_index[field] = idx
                        return True
                    used_indices[idx] = False

        helper(0)


def interpret_data(data: str) -> Tuple[TicketRule, List[int], List[List[int]]]:
    parts: List[str] = data.split('\n\n')
    rule: TicketRule = TicketRule(parts[0])

    my_ticket: str = parts[1].splitlines()[1]
    my_ticket = [int(value) for value in my_ticket.split(',')]

    nearby_tickets: List[str] = parts[2].splitlines()[1:]
    nearby_tickets: List[List[int]] = [
        [int(value) for value in ticket.split(',')] for ticket in nearby_tickets]

    return rule, my_ticket, nearby_tickets


def part_1(rule: TicketRule, nearby_tickets: List[List[int]]) -> int:
    return sum([rule.count_invalid(ticket) for ticket in nearby_tickets])


def part_2(rule: TicketRule, my_ticket: List[int]) -> int:
    rule.detect_index()
    indices = []
    for field in rule.rules:
        if 'departure' in field:
            indices.append(rule.rule_index[field])

    result = 1
    for idx in indices:
        result *= my_ticket[idx]
    return result


class Testing(unittest.TestCase):

    def test_field_range(self):
        rule = FieldRule("1-3 or 5-7")

        self.assertTrue(rule.validate(1))
        self.assertTrue(rule.validate(2))
        self.assertTrue(rule.validate(3))
        self.assertTrue(rule.validate(5))
        self.assertTrue(rule.validate(6))
        self.assertTrue(rule.validate(7))

        self.assertFalse(rule.validate(0))
        self.assertFalse(rule.validate(4))
        self.assertFalse(rule.validate(8))
        self.assertFalse(rule.validate(100))

    def test_part_1(self):
        rule, _, nearby_tickets = interpret_data(test_input)

        self.assertEqual(part_1(rule, nearby_tickets), 71)

    def test_part_2(self):
        rule, my_ticket, nearby_tickets = interpret_data(test_input2)
        part_1(rule, nearby_tickets)
        part_2(rule, my_ticket)
        self.assertEqual(rule.rule_index.get('row'), 0)
        self.assertEqual(rule.rule_index.get('class'), 1)
        self.assertEqual(rule.rule_index.get('seat'), 2)

    def test_part_2_2(self):
        rule, my_ticket, nearby_tickets = interpret_data(test_input)
        part_1(rule, nearby_tickets)
        part_2(rule, my_ticket)
        self.assertEqual(rule.rule_index.get('row'), 0)
        self.assertEqual(rule.rule_index.get('class'), 1)
        self.assertEqual(rule.rule_index.get('seat'), 2)


if __name__ == '__main__':
    print('Day 16')
    data: str = read_input('../inputs/day16.txt')
    rule, my_ticket, nearby_tickets = interpret_data(data)

    print('\tPart 1: {}'.format(part_1(rule, nearby_tickets)))
    print('\tPart 2: {}'.format(part_2(rule, my_ticket)))
    unittest.main()

File: rust/src/day16.rs
use crate::utilities::read_block;
use std::collections::HashMap;

struct FieldRule {
    first_range: (i64, i64),
    second_range: (i64, i64),
}

impl FieldRule {
    fn new(rule: &str) -> Self {
        let mut ranges = rule.split(" or ");
        Self {
            first_range: Self::read_range(ranges.next().unwrap()),
            second_range: Self::read_range(ranges.next().unwrap()),
        }
    }

    fn read_range(range: &str) -> (i64, i64) {
        let mut numbers = range.split('-');
        let low = numbers.next().unwrap().parse().unwrap();
        let high = numbers.next().unwrap().parse().unwrap();
        (low, high)
    }

    fn validate(&self, value: i64) -> bool {
        (self.first_range.0 <= value && self.first_range.1 >= value)
            || (self.second_range.0 <= value && self.second_range.1 >= value)
    }

    fn validate_column(&self, tickets: &Vec<Vec<i64>>, index: usize) -> bool {
        tickets.iter().all(|ticket| self.validate(ticket[index]))
    }
}

struct TicketRule {
    fields: HashMap<String, FieldRule>,
    valid_tickets: Vec<Vec<i64>>,
    field_index: HashMap<String, usize>,
}

impl TicketRule {
    pub fn new(rules: &String) -> Self {
        let mut fields = HashMap::new();
        for rule in rules.split('\n') {
            let mut iter = rule.split(": ");
            let field = iter.next().unwrap().to_string();
            fields.insert(field, FieldRule::new(iter.next().unwrap()));
        }
        Self {
            fields,
            field_index: HashMap::new(),
            valid_tickets: Vec::new(),
        }
    }

    fn validate_value(&self, value: i64) -> bool {
        self.fields.values().any(|field| field.validate(value))
    }

    pub fn sum_invalid(&mut self, nearby_tickets: &String) -> i64 {
        let tickets: Vec<Vec<i64>> = nearby_tickets
            .split('\n')
            .skip(1)
            .map(|line| {
                line.split(',')
                    .map(|n| n.parse::<i64>().unwrap())
                    .collect::<Vec<i64>>()
            })
            .collect();
        let mut ans = 0;
        for ticket in tickets {
            let mut valid = true;
            for &value in &ticket {
                if !self.validate_value(value) {
                    valid = false;
                    ans += value;
                }
            }
            if valid {
                self.valid_tickets.push(ticket);
            }
        }
        ans
    }

    fn calculate_ticket_product(&mut self, my_ticket: &String) -> i64 {
        let my_ticket: Vec<i64> = my_ticket
            .split('\n')
            .skip(1)
            .next()
            .unwrap()
            .split(',')
            .map(|n| n.parse().unwrap())
            .collect();

        let mut field_possible_indices = HashMap::new();
        for (field, rule) in &self.fields {
            for index in 0..my_ticket.len() {
                if rule.validate_column(&self.valid_tickets, index) {
                    field_possible_indices
                        .entry(field.clone())
                        .or_insert(vec![])
                        .push(index);
                }
            }
        }
        let mut fields_by_len: Vec<(usize, String)> = field_possible_indices
            .iter()
            .map(|(field, possible_choices)| (possible_choices.len(), field.clone()))
            .collect();

        fields_by_len.sort();

        let mut available_indices = vec![true; my_ticket.len()];

        self.validate_index(
            &fields_by_len,
            &field_possible_indices,
            &mut available_indices,
        );

        self.field_index
            .iter()
            .filter(|(field, _)| field.starts_with("departure"))
            .map(|(_, &index)| my_ticket[index])
            .fold(1, |acc, x| acc * x)
    }

    fn validate_index(
        &mut self,
        fields_by_len: &[(usize, String)],
        field_possible_indices: &HashMap<String, Vec<usize>>,
        available_indices: &mut Vec<bool>,
    ) -> bool {
        if fields_by_len.is_empty() {
            true
        } else {
            let field = &fields_by_len[0].1;
            for &index in field_possible_indices.get(field).unwrap() {
                if available_indices[index] {
                    available_indices[index] = false;
                    if self.validate_index(
                        &fields_by_len[1..],
                        field_possible_indices,
                        available_indices,
                    ) {
                        self.field_index.insert(field.to_string(), index);
                        return true;
                    }
                    available_indices[index] = true;
                }
            }
            false
        }
    }
}

pub fn main() {
    println!("Day 16");
    let data = read_block("../inputs/day16.txt");
    let mut ticket_rule = TicketRule::new(&data[0]);
    println!("\tPart 1: {}", ticket_rule.sum_invalid(&data[2]));
    println!(
        "\tPart 2: {}",
        ticket_rule.calculate_ticket_product(&data[1])
    );
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_field_range() {
        let rule = FieldRule::new("1-3 or 5-7");
        assert!(rule.validate(1));
        assert!(rule.validate(2));
        assert!(rule.validate(3));
        assert!(rule.validate(5));
        assert!(rule.validate(6));
        assert!(rule.validate(7));

        assert!(!rule.validate(0));
        assert!(!rule.validate(4));
        assert!(!rule.validate(8));
        assert!(!rule.validate(100));
    }

    #[test]
    fn test_sum_invalid() {
        let data = vec![
            "class: 1-3 or 5-7\nrow: 6-11 or 33-44\nseat: 13-40 or 45-50".to_string(),
            "your ticket:\n7,1,14".to_string(),
            "nearby tickets:\n7,3,47\n40,4,50\n55,2,20\n38,6,12".to_string(),
        ];

        let mut ticket_rule = TicketRule::new(&data[0]);
        assert_eq!(ticket_rule.sum_invalid(&data[2]), 71);
    }
}
